### ISSUE TITLE
Dependency resolution performance for large projects 

### ISSUE THREAD

**Issue by moley** [gradle/gradle#6171]
We are facing a performance issue on resolving dependencies with gradle. 
We have a large project which we modularize. At the end we will have about 20 modules, each of them is a multi project build with about 10-20 subprojects. We also modularized our framework, which comes along with about 30 subprojects. We are half the way and have dependency resolution times per module of 30 seconds each time. I have done different performance measurements coming to inconclusive results, but all together they are ways to slow. I have uploaded a testproject to [github](https://github.com/moley/exampleGradleDepsPerformance). Details on handling you can see in README.md. 
For us this is currently a show stopper for this modularization, so it would be nice to get a solution, workaround or a tip what goes wrong in our scenario asap. If you have problems to reproduce or any further questions, please call. 
I have also opened a bug in [spring-gradle-plugin,](https://github.com/spring-projects/spring-boot/issues/13979) because using this plugin makes the performance even worse.
Thanks in advance

Cheers
Markus

**Comment by oehme** [comment]
@bigdaz All the CPU samples point to `GradlePomModuleDescriptorParser`, probably because of the heavy use of `mavenLocal`. 

@moley Have you considered using composite builds instead of mavenLocal? That would give you much faster turnarounds than installing to mavenLocal.

**Comment by bigdaz** [comment]
Thanks @moley for the sample project. 

Since Gradle bypasses the cache for file-backed local repositories, we are aware that this will introduce some overhead. However, it's good to get some real data on just how big of a pain this is. 

In theory we should be only parsing each Maven POM once per build invocation, but perhaps there's a bug that's preventing this from working. Your sample should help us investigate this issue.

**Comment by moley** [comment]
@bigdaz No problem, you're welcome

@oehme I always wanted to work with composite build but had no time so far to learn about it :-) Perhaps this is my chance. But  - even if we don't use mavenLocal the performance is poor. Using mavenLocal so often is only an issue in the context of this testproject and in the migration phase Afterwards I would estimate we need mavenLocal, but far not so often. I could change my test project to work with composite builds in another branch, so you can still investigate the mavenLocal performance and in parallel the other problem.... Or should I wait until you have finished this mavenLocal thing?

**Comment by bigdaz** [comment]
> even if we don't use mavenLocal the performance is poor

We suspect the performance issue is due to local file repositories, not specifically `mavenLocal`. Are you able to easily test using a local HTTP repository? It would be interesting to see if this confirms our suspicion.

**Comment by moley** [comment]
ok, thank you, I will retry with our real world project and give feedback on this

**Comment by moley** [comment]
@bigdaz simply changing order of repos did not improve things, I think I have to change the testproject to use local http repo

**Comment by moley** [comment]
Today I set up a local artifactory, which improves resolution dependency to 3-4 seconds. Changed the README.md to describe how to reproduce. So it is a workaround for us to use real dependencies or composite builds, but neverltheless would be nice on long term when this issue could be fixed

**Comment by oehme** [comment]
3-4s is still pretty long, definitely worth profiling as well. Thank you for the example!

**Comment by oehme** [comment]
I'm going to close this as a duplicate of https://github.com/gradle/gradle/issues/4322
